src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp

```diff
@@ -335,6 +390,179 @@ void CompositionalMultiphaseWell::UpdateComponentFraction( WellElementSubRegion
+  arrayView1d< real64 > const & dCurrentPhaseVolRate_dRate =
+    wellControls.getReference< array1d< real64 > >( CompositionalMultiphaseWell::viewKeyStruct::dCurrentPhaseVolRate_dRateString );
+
+  real64 & currentTotalVolRate =
+    wellControls.getReference< real64 >( CompositionalMultiphaseWell::viewKeyStruct::currentTotalVolRateString );
+  real64 & dCurrentTotalVolRate_dPres =
+    wellControls.getReference< real64 >( CompositionalMultiphaseWell::viewKeyStruct::dCurrentTotalVolRate_dPresString );
+  arrayView1d< real64 > const & dCurrentTotalVolRate_dCompDens =
+    wellControls.getReference< array1d< real64 > >( CompositionalMultiphaseWell::viewKeyStruct::dCurrentTotalVolRate_dCompDensString );
+  real64 & dCurrentTotalVolRate_dRate =
+    wellControls.getReference< real64 >( CompositionalMultiphaseWell::viewKeyStruct::dCurrentTotalVolRate_dRateString );
+
+
+  // 1) Update the fluid properties in the top well element using the reference pressure
+  //    The phase densities and phase fractions in the top well element will be recomputed with the correct pressure in UpdateFluidModel
+
+  constitutive::constitutiveUpdatePassThru( fluid, [&] ( auto & castedFluid )
+  {
+    typename TYPEOFREF( castedFluid ) ::KernelWrapper fluidWrapper = castedFluid.createKernelWrapper();
+
+    // TODO: send surface pressure and surface temperature here
+    CompositionalMultiphaseWellKernels::FluidUpdateAtReferenceConditionsKernel::Launch< serialPolicy >( targetSet,
+                                                                                                        fluidWrapper,
+                                                                                                        targetBHP,
+                                                                                                        m_temperature,
+                                                                                                        compFrac );
+  } );
+
+  // 2) Use the updated phase densities and phase fractions to compute current BHP and the volumetric rates
+
+  arrayView2d< real64 const > const & compDens =
+    subRegion.getReference< array2d< real64 > >( viewKeyStruct::globalCompDensityString );
+  arrayView2d< real64 const > const & dCompDens =
+    subRegion.getReference< array2d< real64 > >( viewKeyStruct::deltaGlobalCompDensityString );
+
+  std::cout << "" << std::endl;
+  std::cout << subRegion.getName() << std::endl;
+  for( localIndex i = 0; i < subRegion.size(); ++i )
+  {
+    std::cout << "wellElemPressure[" << i << "] = " << pres[i] + dPres[i]
+              << " connRate[" << i << "] = " << connRate[i] + dConnRate[i]
+              << " compDens[" << i << "][0] = " << compDens[i][0] + dCompDens[i][0]
+              << std::endl;
+  }
+
+  // TODO: --------> pass by reference ????
+  forAll< parallelDevicePolicy<> >( targetSet.size(), [&] GEOSX_HOST_DEVICE ( localIndex const a )
+  {
+    localIndex const k = targetSet[a];
```

Comment:
You can also just
```suggestion
  forAll< parallelDevicePolicy<> >( 1, [=] GEOSX_HOST_DEVICE ( localIndex const )
  {
    localIndex const k = iwelemTop;
```
(or use `iwelemTop` directly instead of `k`)

Reply:
Thanks for the quick feedback, I will do that. 

In this piece of code I update some dynamic values (say, `currentBHP`) that will be assembled on the GPU later (for the BHP constraint: `currentBHP - targetBHP`). This update is always done in one well element only. 

So if I do the update on the GPU inside a RAJA `forAll< parallelDevicePolicy<> >` loop and pass everything by value with `[=]`, how should I update a `real64` like the `currentBHP`? With a reduce? Or an `arrayView1d` of size 1? I was unsure about what to do here.

Reply:
Hmm. I did not catch that there's a scalar value update. Indeed we have two options here: either register and capture the scalar as an array of size 1; or run the kernel with `serialPolicy` and use explicit capture to selectively capture scalars by-referece (i.e. `[pres, dPres, ..., &currentBHP]`). I'm leaning towards the latter, although explicit capture lists tend to get super long, but they seem to be the preferred style these days. There will be overhead either way: launching a 1-point GPU kernel, or moving some (not very large) arrays back to host. I don't know which one's worse, but it probably doesn't make a huge difference anyway.

Reply:
@klevzoff Thanks for the suggestions. I implemented the second option that you suggested, with `serialPolicy` and explicit capture of the scalars by reference.